grouper/services/group.py

```diff
@@ -0,0 +1,58 @@
+from typing import TYPE_CHECKING
+
+from grouper.constants import (
+    DEFAULT_ADMIN_GROUP,
+    GROUP_ADMIN,
+    PERMISSION_ADMIN,
+    PERMISSION_AUDITOR,
+    USER_ADMIN,
+)
+from grouper.entities.group import GroupJoinPolicy
+from grouper.usecases.interfaces import GroupInterface
+
+if TYPE_CHECKING:
+    from grouper.repositories.interfaces import GroupRepository, PermissionGrantRepository
+    from grouper.settings import Settings
+
+
+class GroupService(GroupInterface):
+    """High-level logic to manipulate groups."""
+
+    def __init__(self, settings, group_repository, permission_grant_repository):
+        # type: (Settings, GroupRepository, PermissionGrantRepository) -> None
+        self.settings = settings
+        self.group_repository = group_repository
+        self.permission_grant_repository = permission_grant_repository
+
+    def create_group(self, name, description, join_policy):
+        # type: (str, str, GroupJoinPolicy) -> None
+        self.group_repository.create_group(name, description, join_policy)
+
+    def grant_permission_to_group(self, permission, argument, group):
+        # type: (str, str, str) -> None
+        self.permission_grant_repository.grant_permission_to_group(permission, argument, group)
+
+    def group_exists(self, name):
+        # type: (str) -> bool
+        return True if self.group_repository.get_group(name) else False
+
+    def initialize_administrator_group(self):
```

Comment:
For my own understanding, what's the reasoning behind putting the logic here (and in `initialize_auditors_group`) in the service instead of being in the usecase? As an eye test it looks like a bit of actual logic that consists of calling a bunch of service functions. Guessing there's also only a single usecase that will end up calling this.

Reply:
Yeah, in retrospect, my justification for doing this (not wanting to plumb settings into the usecase) was a bad one. I'll change this.